### PR TITLE
[stdlib]String comparison

### PR DESCRIPTION
**Pull request by lancep**
<rdar://problem/18992510>

### CODE REVIEW

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
CC @milseman

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke test Linux platform

**Review by milseman — COMMENTED** [review]
Are you familiar with the test cases in `StringComparisonTest.allTests`, and can we keep test coverage? We could loop over them, and make sure the "force-normalize-and-compare" path is equivalent to comparison.

**Comment by lancep** [comment]
@milseman I agree we should keep them, once I get the tests passing I'll re-introduce the file and try to save them.

**Comment by lancep** [comment]
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1393

@swift-ci Please smoke test Linux platform

**Comment by lancep** [comment]
@swift-ci please smoke test OS X platform

**Comment by glessard** [comment]
Thanks for this! It's nice that these bugs and inconsistencies will finally be resolved

**Comment by milseman** [comment]
@lancep we need to investigate the ASCII regressions before landing. It might just be that code got a little perturbed and they're fragile micro benchmarks, but worth a look.

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1393

@swift-ci Please test Linux platform

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
This is ready to go other than the ascii benchmark regression which I'm looking into now.

File: stdlib/public/core/FixedArray.swift.gyb [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 %{
   # The sizes to generate code for.
-  sizes = [16]
+  sizes = [2, 8, 16]
```
Comment (milseman) [inline stdlib/public/core/FixedArray.swift.gyb]: Are you still using all these sizes?

Comment (lancep) [reply stdlib/public/core/FixedArray.swift.gyb]: nope, we can get rid of them!

Comment (lancep) [reply stdlib/public/core/FixedArray.swift.gyb]: actually not, we are still using them

**Comment by milseman** [comment]
These are not acceptable regressions for `StringEqualPointerComparison`, `Phonebook `, `Dictionary`, etc., as well. The good news is that they're likely to be the same ARC overhead as `StringComparison_ascii`.

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test OS X platform

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test Linux platform

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@lancep let me know when you've done your review of the diff and cleanup, and I'll review

**Comment by milseman** [comment]
@swift-ci please test compiler performance

File: stdlib/public/core/StringComparison.swift [hunk 2]
```diff
@@ -298,7 +294,7 @@ internal func _lexicographicalCompare(
   let rightHSPtr = rightHS.baseAddress._unsafelyUnwrappedUnchecked
   return _lexicographicalCompare(leftHSPtr[idx], rightHSPtr[idx])
 }
-// @inline(never) // TODO: temporary, to inspect code quality
+@inline(__always)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Why do you think this should always be inlined?

File: stdlib/public/core/StringComparison.swift [hunk 3]
```diff
@@ -283,7 +279,7 @@ internal func _lexicographicalCompare(
 ) -> _Ordering {
   return lhs < rhs ? .less : (lhs > rhs ? .greater : .equal)
 }
-// @inline(never) // TODO: temporary, to inspect code quality
+@inline(__always)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Why do you think this should always be inlined?

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringComparison.swift (hunk 2 above)
Comment (lancep) [reply stdlib/public/core/StringComparison.swift]: This matches the other lexicographicalCompare overloads in this file

File: stdlib/public/core/StringComparison.swift (hunk 3 above)
Comment (lancep) [reply stdlib/public/core/StringComparison.swift]: There are other overloads of _lexicographicalCompare that were marked that way, so I assumed that's what you had there before you made it @inline(never) with a TODO.

File: stdlib/public/core/FixedArray.swift.gyb [hunk 4]
```diff
@@ -56,28 +56,25 @@ extension _FixedArray${N} {
...
     @inline(__always) set { _count = Int8(newValue) }
   }
 }
 
 extension _FixedArray${N} : RandomAccessCollection, MutableCollection {
   internal typealias Index = Int
+  internal typealias IndexDistance = Int
 
-  @_inlineable // FIXME(sil-serialize-all)
-  @_versioned // FIXME(sil-serialize-all)
```
[the remark this answers is not in the document]
Comment (lancep) [reply stdlib/public/core/FixedArray.swift.gyb]: I undid this already, see the updated diff

File: stdlib/public/core/ICU.swift [hunk 5]
```diff
@@ -22,9 +22,9 @@ extension __swift_stdlib_UErrorCode {
   internal var isWarning: Bool {
     return rawValue < __swift_stdlib_U_ZERO_ERROR.rawValue
   }
-  @_inlineable // FIXME(sil-serialize-all)
-  @_versioned // FIXME(sil-serialize-all)
-  internal var isSuccess: Bool {
+  // @_inlineable // FIXME(sil-serialize-all)
+  // @_versioned // FIXME(sil-serialize-all)
```
[the remark this answers is not in the document]
Comment (lancep) [reply stdlib/public/core/ICU.swift]: ditto

File: stdlib/public/core/StringComparable.swift [hunk 6]
```diff
@@ -211,22 +77,18 @@ extension _StringGuts {
...
       let rightASCII = right._unmanagedASCIIView[rightRange]
       let result = leftASCII.compareASCII(to: rightASCII)
       return result
     }
-#endif
-    return _compareDeterministicUnicodeCollation(
-      _leftUnsafeStringGutsBitPattern: left.rawBits, leftRange,
-      _rightUnsafeStringGutsBitPattern: right.rawBits, rightRange)
+    
+    let selfBits = _explodeBits(left)
```
Comment (milseman) [inline stdlib/public/core/StringComparable.swift]: This already exists. See `_StringGuts.rawBits` and `_StringGuts.init(rawBits:)`, and use those for consistency.

File: stdlib/public/core/StringComparison.swift [hunk 7]
```diff
@@ -0,0 +1,1368 @@
...
+ func _implodeBits(_ bits: _StringGuts._RawBitPattern) -> _StringGuts {
+   return _StringGuts(
+     object: Builtin.reinterpretCast(bits.0),
+     otherBits: Builtin.reinterpretCast(bits.1))
+  }
+
+@inline(never)
+@effects(readonly)
+public
+func _compareHack(
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: What is the hack? Why is it there? Please provide some annotation or otherwise. Also, comparison was already hacking around ARC, so how is this approach different?

File: stdlib/public/core/StringComparison.swift [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+//   var rebased: UnsafeMutableBufferPointer<UInt16> {
+//     return UnsafeMutableBufferPointer(rebasing: self)
+//   }
+// }
+
+// extension Slice where Base == UnsafeMutableBufferPointer<Base.Element> {
+//   var rebased: UnsafeMutableBUfferPointer<Base.Element> {
+//     return UnsafeMutableBufferPointer(rebasing: self)
+//   }
+// }
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Let's drop the commented out code.

File: stdlib/public/core/StringComparison.swift [hunk 9]
```diff
@@ -0,0 +1,1368 @@
...
+  _sanityCheck(_isLeadingSurrogate(high) && _isTrailingSurrogate(low))
+  let hi10: UInt32 = UInt32(high) &- UInt32(_leadingSurrogateBias)
+  _sanityCheck(hi10 < 1<<10, "I said high 10. Not high, like, 20 or something")
+  let lo10: UInt32 = UInt32(low) &- UInt32(_trailingSurrogateBias)
+  _sanityCheck(lo10 < 1<<10, "I said low 10. Not low, like, 20 or something")
+
+  let result = ((hi10 &<< 10) | lo10) &+ 0x1_00_00
+  // print("""
+  //   hi10: \(hex(hi10)), lo10: \(hex(lo10)) ==> \(hex(result))
+  // """)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Please remove commented out code. If you want to keep this kind of thing around for future debugging purposes, then we should come up with a different scheme for that.

File: stdlib/public/core/StringComparison.swift [hunk 10]
```diff
@@ -0,0 +1,1368 @@
...
+  // TODO: inspect code quality
+  return lhs < rhs ? .less : (lhs > rhs ? .greater : .equal)
+}
+@inline(__always)
+internal func _lexicographicalCompare(
+  _ lhs: UInt16, _ rhs: UInt16
+) -> _Ordering {
+  return lhs < rhs ? .less : (lhs > rhs ? .greater : .equal)
+}
+@inline(__always)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: I don't think you want to always inline this

File: stdlib/public/core/StringComparison.swift [hunk 11]
```diff
@@ -0,0 +1,1368 @@
...
+
+  let idx = _findDiffIdx(leftHS, rightHS)
+  guard idx < count else {
+    return _lexicographicalCompare(leftHS.count, rightHS.count)
+  }
+  let leftHSPtr = leftHS.baseAddress._unsafelyUnwrappedUnchecked
+  let rightHSPtr = rightHS.baseAddress._unsafelyUnwrappedUnchecked
+  return _lexicographicalCompare(leftHSPtr[idx], rightHSPtr[idx])
+}
+@inline(__always)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Same here

File: stdlib/public/core/StringComparison.swift [hunk 12]
```diff
@@ -0,0 +1,1368 @@
...
+  }
+  let nextCu: UInt16 = ptr[idx+1]
+  guard _isTrailingSurrogate(nextCu) else {
+    // Invalid surrogate pair: just return the invalid value
+    return (UnicodeScalar(_unchecked: UInt32(cu)), idx+1)
+  }
+
+  // print("""
+  //   leading: \(hex(cu)), trailing: \(hex(nextCu))
+  // """)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Delete commented out code

File: stdlib/public/core/StringComparison.swift [hunk 13]
```diff
@@ -0,0 +1,1368 @@
...
+  _sanityCheck(Int32(exactly: value) != nil, "top bit shouldn't be set")
+  return (UnicodeScalar(_unchecked: value), idx+2)
+}
+internal func _parseRawScalar(
+  _ guts: _StringGuts,
+  startingFrom idx: Int = 0
+) -> (UnicodeScalar, scalarEndIndex: Int) {
+  var buffer = _FixedArray2<UInt16>(allZeros:())
+  buffer[0] = guts[idx]
+  buffer[1] = guts[idx+1]
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: This is a bad idea and _StringGuts should remove this subscript. We should branch once at the top on the representation, and not repeatedly inside our tight loops. Why is this needed?

File: stdlib/public/core/StringComparison.swift [hunk 14]
```diff
@@ -0,0 +1,1368 @@
...
+    // Otherwise, we need to grow
+    guard count <= canary else {
+      fatalError("Invariant broken: Max decomposition factor insufficient")
+    }
+    count *= 2
+  }
+}
+
+extension _UnmanagedString where CodeUnit == UInt8 {
+  /// This is consistent with Foundation, but incorrect as defined by Unicode.
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: This comment is not true or relevant anymore. Please drop and/or replace

File: stdlib/public/core/StringComparison.swift [hunk 15]
```diff
@@ -0,0 +1,1368 @@
...
+  @_versioned
+  internal
+  func _unsafeUTF16(_ range: Range<Int>) -> UnsafeBufferPointer<UInt16> {
+    _sanityCheck(!isASCII)
+    return UnsafeBufferPointer(
+      start: _unmanagedUTF16View[range].start,
+      count: range.count
+    )
+  }
+
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: These already exist. Why not use the _unmanaged views already on _StringGuts?

File: stdlib/public/core/StringComparison.swift [hunk 16]
```diff
@@ -0,0 +1,1368 @@
...
+  public
+  func _compareOpaque(_ selfRange: Range<Int>, _ other: _StringGuts, _ otherRange: Range<Int>) -> _Ordering {
+    //
+    // Do a fast Latiny comparison loop; bail if that proves insufficient.
+    //
+    // The vast majority of the time, seemingly-non-contiguous Strings are
+    // really ASCII strings that were bridged improperly. E.g., unknown nul-
+    // termination of an all-ASCII file loaded by String.init(contentsOfFile:).
+    //
+
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Shouldn't this be expressed on `_UnmanagedOpaqueString`, and not `_StringGuts`?

File: stdlib/public/core/StringComparison.swift [hunk 17]
```diff
@@ -0,0 +1,1368 @@
...
+  }
+  if _slowPath(idx+1 == buffer.count) {
+    return true
+  }
+
+  let nextCU = ptr[idx+1]
+  return nextCU < 0x300 || _hasNormalizationBoundary(before: nextCU)
+}
+
+public func _compareStringsPreLoop(
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: All of these global functions can now be on the relevant `_UnmanagedString where CodeUnit == UIntX` structures.

File: stdlib/public/core/StringComparison.swift [hunk 18]
```diff
@@ -0,0 +1,1368 @@
...
+        buffer[bufferIndex] = cu
+        index += 1
+        bufferIndex += 1
+      } while !guts.hasNormalizationBoundary(after: index - 1)
+      
+      return bufferIndex
+    }
+  }
+  
+  struct BufferPointerSource: SegmentSource {
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Again, looks like you want `UnmanagedString` and `UnmanagedOpaqueString`.

File: stdlib/public/core/StringComparison.swift [hunk 19]
```diff
@@ -0,0 +1,1368 @@
...
+  mutating func tryFill(buffer: UnsafeMutableBufferPointer<UInt16>) -> Int?
+}
+
+extension SegmentSource {
+  mutating func tryFill(buffer: UnsafeMutablePointer<_Normalization._SegmentOutputBuffer>) -> Int? {
+    return tryFill(buffer: _castOutputBuffer(buffer))
+  }
+}
+
+struct NormalizedIterator: IteratorProtocol {
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Extract into a separate source file.

File: stdlib/public/core/StringComparison.swift [hunk 20]
```diff
@@ -0,0 +1,1368 @@
...
+  // Fast path: sub-0x300 single-scalar segments can be compared directly
+  //
+  if _fastPath(
+    _isLatinyPrenormal(selfUTF16, idx: idx)
+    && _isLatinyPrenormal(otherUTF16, idx: idx)
+  ) {
+    return _lexicographicalCompare(selfCU, otherCU)
+  }
+
+  // print("""
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Delete commented out code

File: stdlib/public/core/StringComparison.swift [hunk 21]
```diff
@@ -0,0 +1,1368 @@
...
+      otherUTF16: UnsafeBufferPointer<UInt16>(rebasing: otherUTF16[comparisonStartIdx...])
+      )
+  }
+  let selfLength = selfSegmentLengthOpt._unsafelyUnwrappedUnchecked
+  let otherLength = otherSegmentLengthOpt._unsafelyUnwrappedUnchecked
+
+  // print("""
+  //   comparing \(selfOutputBuffer.hex)[..<\(selfLength)] \
+  //   to \(otherOutputBuffer.hex)[..<\(otherLength)]
+  //   """)
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Delete commented out code

File: stdlib/public/core/StringComparison.swift [hunk 22]
```diff
@@ -0,0 +1,1368 @@
...
+  // lasting correctness, because Unicode. (Also, sometimes it's impossible, but
+  // that's what _compareStringsPathological is for.) However, it helps for very
+  // long strings that differ in the middle. We might want this one day... but
+  // not today.
+  //
+  // TODO: An additional (or even repeated) reapplying of the algorithm,
+  // including the binary diff scan, could greatly benefit strings that only
+  // sparsely differ in normality (penalizing strings that densely differ in
+  // normality). This would add complexity, but with compelling data could be an
+  // alternative to chunking.
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Did you ever see this kind of scenario in benchmarking?

File: stdlib/public/core/StringNormalization.swift [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+      Int32(buffer.count),
+      &err)
+
+    guard err.isSuccess else {
+      // This shouldn't be possible unless some deep (unrecoverable) system
+      // invariants are violated
+      fatalError("Unable to talk to ICU")
+    }
+    return length == buffer.count
+  }}
```
Comment (milseman) [inline stdlib/public/core/StringNormalization.swift]: missing newline

File: stdlib/public/core/StringNormalization.swift [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+  // segment. Fits all but pathological arbitrary-length segments (i.e. zalgo-
+  // segments)
+  internal typealias _SegmentOutputBuffer = _FixedArray16<UInt16>
+
+  // // A large output buffer to use for normalizing large portions of a String. It
+  // // should be used extremely carefully, as chunking is inherently tricky in
+  // // Unicode, even when aligned on a normalization boundary (see "Multi-Segment
+  // // Expanders" below)
+  // public // String Comparison Prototype
+  // typealias _ChunkOutputBuffer = _FixedArray64<UInt16>
```
Comment (milseman) [inline stdlib/public/core/StringNormalization.swift]: Remove commented out code

File: stdlib/public/stubs/UnicodeNormalization.cpp [hunk 25]
```diff
@@ -151,9 +169,9 @@ swift::_swift_stdlib_unicode_compare_utf16_utf16(const uint16_t *LeftString,
   // as wchar_t which size is 2. It seems that the underlying binary
   // representation is same with swift utf16 representation.
   // On Clang 4.0 under a recent Linux, ICU uses the built-in char16_t type.
```
Comment (milseman) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: Can these all be removed?

File: stdlib/public/stubs/UnicodeNormalization.cpp [hunk 26]
```diff
@@ -214,7 +237,8 @@ void *swift::_swift_stdlib_unicodeCollationIterator_create(
       ucol_openElements(GetRootCollator(), reinterpret_cast<const UChar *>(Str),
                         Length, &ErrorCode);
   if (U_FAILURE(ErrorCode)) {
-    swift::crash("_swift_stdlib_unicodeCollationIterator_create: ucol_openElements() failed.");
+    swift::crash("_swift_stdlib_unicodeCollationIterator_create: "
```
Comment (milseman) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: All this collation code can be removed, right?

File: test/SILOptimizer/string_switch.swift [hunk 27]
```diff
@@ -1,6 +1,7 @@
 // RUN: %target-build-swift -O %s -module-name=test -Xllvm -sil-disable-pass=FunctionSignatureOpts -o %t.out
 // RUN: %target-build-swift -O %s -module-name=test -Xllvm -sil-disable-pass=FunctionSignatureOpts -emit-sil | %FileCheck %s
 // RUN: %target-run %t.out
+// REQUIRES: swift_stdlib_no_asserts,optimized_stdlib
```
Comment (milseman) [inline test/SILOptimizer/string_switch.swift]: Is this still necessary?

File: validation-test/stdlib/String.swift [hunk 28]
```diff
@@ -1937,4 +1919,119 @@ for test in testCases {
...
+  // var test: (String, String) -> Void
+  var comparison: Comparison
+  
+  init(_ strings: [String], _ comparison: Comparison) {
+    self.strings = strings
+    self.comparison = comparison
+  }
+  
+  func test() {
+    for pair in zip(strings, strings[1...]) {
```
Comment (milseman) [inline validation-test/stdlib/String.swift]: Should we test both the main API path as well as the normalized iterator path?

**Review by milseman — CHANGES_REQUESTED** [review]

File: test/SILOptimizer/string_switch.swift (hunk 27 above)
Comment (lancep) [reply test/SILOptimizer/string_switch.swift]: I'm not sure, I'll make sure the tests pass with this, then I can try taking it out again.

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke benchmark

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke test Linux platform

File: stdlib/public/stubs/UnicodeNormalization.cpp (hunk 26 above)
Comment (lancep) [reply stdlib/public/stubs/UnicodeNormalization.cpp]: String hashing still uses some of these functions on Linux

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke test OS X platform

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci Please smoke benchmark

File: stdlib/public/core/NormalizedCodeUnitIterator.swift [hunk 29]
```diff
@@ -0,0 +1,253 @@
...
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+internal
+struct NormalizedCodeUnitIterator: IteratorProtocol {
```
Comment (milseman) [inline stdlib/public/core/NormalizedCodeUnitIterator.swift]: If this becomes @_versioned, I would say it's better to underscore the name so that we reserve this name for the future. Might as well do it now.

File: stdlib/public/core/NormalizedCodeUnitIterator.swift [hunk 30]
```diff
@@ -0,0 +1,253 @@
...
+  var segmentBuffer = _FixedArray16<CodeUnit>(allZeros:())
+  var overflowBuffer: [CodeUnit]? = nil
+  var normalizationBuffer: [CodeUnit]? = nil
+  var source: SegmentSource
+  var segmentBufferIndex = 0
+  var segmentBufferCount = 0
+  var overflowBufferIndex = 0
+  var overflowBufferCount = 0
```
Comment (milseman) [inline stdlib/public/core/NormalizedCodeUnitIterator.swift]: I think this is ok for now. I hope to build up more low-level structures and facilities to do this book-keeping for us without significant overhead.

File: stdlib/public/core/NormalizedCodeUnitIterator.swift [hunk 31]
```diff
@@ -0,0 +1,253 @@
...
+        bufferIndex += 1
+      } while unmanagedString.hasNormalizationBoundary(
+          after: index - 1, 
+          count: unmanagedString.count) == false
+      
+      return bufferIndex
+    }
+  }
+  
+  struct StringGutsSource: SegmentSource {
```
Comment (milseman) [inline stdlib/public/core/NormalizedCodeUnitIterator.swift]: Probably underscore the name.

I think we'll adjust this approach later this release, either in favor of just unmanaged opaque string or a more generally useful streaming opaque string, but that's pending some other work.

File: stdlib/public/core/NormalizedCodeUnitIterator.swift [hunk 32]
```diff
@@ -0,0 +1,253 @@
...
+    
+    if overflowBufferCount == overflowBufferIndex {
+      overflowBufferCount = 0
+      overflowBufferIndex = 0
+    }
+    
+    if source.remaining <= 0 
+    && segmentBufferCount == 0 
+    && overflowBufferCount == 0 {
+      //Our source of code units to normalize is empty and our buffers from
```
Comment (milseman) [inline stdlib/public/core/NormalizedCodeUnitIterator.swift]: General formatting, we always have a space after the `//` for comments

File: stdlib/public/core/NormalizedCodeUnitIterator.swift [hunk 33]
```diff
@@ -0,0 +1,253 @@
...
+      //previous normalizations are also empty.
+      return nil
+    }
+    
+    if segmentBufferCount == 0 && overflowBufferCount == 0 {
+      //time to fill a buffer if possible. Otherwise we are done, return nil
+      // Normalize segment, and then compare first code unit
+      var intermediateBuffer = _FixedArray16<CodeUnit>(allZeros:())
+      if overflowBuffer == nil, let filled = source.tryFill(buffer: &intermediateBuffer) {
+        guard let count = _tryNormalize(_castOutputBuffer(&intermediateBuffer, endingAt: filled), into: &segmentBuffer) else {
```
Comment (milseman) [inline stdlib/public/core/NormalizedCodeUnitIterator.swift]: Have you checked 80-col?

File: stdlib/public/core/StringNormalization.swift [hunk 34]
```diff
@@ -0,0 +1,125 @@
...
+  // musical notes). This expansion is capped by the maximum expansion factor of
+  // the normal form. For NFC, that is 3x.
+  internal static let _maxNFCExpansionFactor = 3
+
+  // A small output buffer to use for normalizing a single normalization
+  // segment. Fits all but pathological arbitrary-length segments (i.e. zalgo-
+  // segments)
+  internal typealias _SegmentOutputBuffer = _FixedArray16<UInt16>
+
+  // // A large output buffer to use for normalizing large portions of a String. It
```
Comment (milseman) [inline stdlib/public/core/StringNormalization.swift]: Dead comment

File: stdlib/public/core/UnmanagedString.swift [hunk 35]
```diff
@@ -173,6 +173,36 @@ extension _UnmanagedString : _StringVariant {
...
+  
+  @_inlineable // FIXME(sil-serialize-all)
+  @_versioned // FIXME(sil-serialize-all)
+  internal subscript(offsetRange: PartialRangeThrough<Int>) -> SubSequence {
+    _sanityCheck(offsetRange.upperBound < count)
+    return _UnmanagedString(
+      start: start, 
+      count: offsetRange.upperBound + 1
+    )
+  }
```
Comment (milseman) [inline stdlib/public/core/UnmanagedString.swift]: Woo!

File: validation-test/stdlib/String.swift [hunk 36]
```diff
@@ -1937,4 +1928,136 @@ for test in testCases {
...
+        expectGT(pair.0, pair.1)
+      case .equal:
+        expectEqual(pair.0, pair.1)
+      }
+    }
+  }
+  
+  func testOpaquePath() {
+#if _runtime(_ObjC)
+    let opaqueStrings = strings.map { NSSlowString(string: $0) as String }
```
Comment (milseman) [inline validation-test/stdlib/String.swift]: 👍

**Review by milseman — APPROVED** [review]
Minor cosmetic tweaks recommended, after which this is LGTM. Thank you!

File: stdlib/public/core/StringComparison.swift [hunk 37]
```diff
@@ -0,0 +1,1164 @@
...
+  mutating func fill(from other: _UnmanagedString<T>) {
+    _sanityCheck(other.count < _FixedArray16<T>.capacity,
+      "out of bounds fill")
+    for i in 0..<other.count {
+      self[i] = other[i]
+    }
+  }
+}
+
+public enum _Ordering: Int, Equatable {
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Make as much as you can `@_versioned internal` instead of public. Public may be need for testing

**Review by milseman — COMMENTED** [review]
(More minor comments)

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by lancep** [comment]
Please test with following pull request:
apple/swift-corelibs-foundation#1393

@swift-ci please smoke test

**Comment by milseman** [comment]
<rdar://problem/18992510>
